src/workerd/io/io-context.c++

```diff
@@ -925,22 +925,52 @@ jsg::AsyncContextFrame::StorageScope IoContext::makeAsyncTraceScope(Worker::Lock
       js, lock.getTraceAsyncContextKey(), js.v8Ref(spanHandle));
 }
 
-SpanBuilder IoContext::makeTraceSpan(kj::StringPtr operationName) {
+SpanParent IoContext::getCurrentTraceSpan() {
   // If called while lock is held, try to use the trace info stored in the async context.
   KJ_IF_MAYBE (lock, currentLock) {
     KJ_IF_MAYBE (frame, jsg::AsyncContextFrame::current(*lock)) {
       KJ_IF_MAYBE (value, frame->get(lock->getTraceAsyncContextKey())) {
         auto handle = value->getHandle(*lock);
         jsg::Lock& js = *lock;
         auto& spanParent = jsg::unwrapOpaqueRef<IoOwn<SpanParent>>(js.v8Isolate, handle);
-        return spanParent->newChild(operationName);
+        return spanParent->addRef();
       }
     }
   }
 
   // If async context is unavailable (unset, or JS lock is not held), fall back to heuristic of
   // using the trace info from the most recent active request.
-  return getMetrics().getSpan().newChild(operationName);
+  return getMetrics().getSpan();
+}
+
+void IoContext::setCurrentTraceSpan(jsg::Lock& js, SpanParent spanParent) {
+  // TODO(cleanup): Here we assume that because we are passed a lock reference that we hold the
+  // current lock.  Is there a preferred/better way to go from a jsg::Lock to a Worker::Lock?
+  auto& lock = KJ_REQUIRE_NONNULL(currentLock);
+  KJ_IF_MAYBE (frame, jsg::AsyncContextFrame::current(lock)) {
+    KJ_IF_MAYBE (value, frame->get(lock.getTraceAsyncContextKey())) {
+      // TODO(now): Here we set the trace information by mutating the value that the async context
+      // is pointing to.  This is sufficient to set the trace information used for synchronously
+      // created spans, but it has the side effect of also changing other async contexts using the
+      // same value.  We probably want to mutate the context itself, but AsyncContextFrame does
+      // not currently provide a set() operation?
```

Comment:
> ... AsyncContextFrame does not currently provide a set() operation?

That's definitely intentional. The frame is expected to be immutable once created.

Reply:
Ah, OK...  It seemed like it might have been intentional, since get/set operations seemed like they would have been more straightforward.  But I wasn't sure, and I wasn't able to readily identify where the immutability requirement was coming from.

Hmm... I might be able to do something more like `asyncLocalStorage.run(state, func)`, instead.  In which case, maybe I can get away without changes to `IoContext`, after all.

Reply:
Look for `jsg::AsyncContextFrame::StorageScope` .. it does what you want.

Reply:
Thanks!  I think I've figured out an approach that uses `StorageScope` and behaves how I'd expect for async operations... I've updated the PR and rebased.